### DISCUSSION TITLE
reshape()

### DISCUSSION THREAD

**Discussion by patgrasso** [josdejong/mathjs#2713] [Ideas]
I think it might be useful to have a `reshape()` function that accepts a matrix and a list of dimensions, and formats the matrix to fit the specified dimensions. `resize()` does this already, but it doesn't preserve all of the elements.

``` javascript
var A = [1, 2, 3, 4, 5, 6, 7, 8, 9];
A.resize([3, 3]);
/* => [[1, 0, 0],
       [2, 0, 0],
       [3, 0, 0]] */
A.reshape([3, 3]);
/* => [[1, 2, 3],
       [4, 5, 6],
       [7, 8, 9]] */
```

Thoughts? I can start working on it and submit a PR if this sounds appealing.

**Comment by josdejong** [comment]
A function `reshape` is a common utility and would be complementary to `resize`. I personally haven't seen any use cases for it though: normally the rows and columns of a matrix containing data have a meaning. Like storing 3d points in rows, where columns 0, 1, 2 hold the x, y, and z location of the points. "Reshuffling" the data points using a function `reshape` simply turns the data in a bunch of meaningless numbers. 

Do you have any practical use cases for `reshape`?

**Comment by patgrasso** [comment]
Primarily as a facilitation when generating data to work with.
For instance, reshaping a range or random data:

``` javascript
var A = math.range(1, 10).reshape([3, 3]);
var B = math.random(0, 1).reshape([2, 3]);
// this is equal to math.random([2, 3]);
```

Or perhaps even adding this functionality to `range()`, as it exists for `random()`, however if for some reason you wanted to use the function outside of these two instances, you wouldn't be able to without the standalone `reshape()` function.

**Comment by ericman314** [comment]
In the past I have used matlab's `reshape` function to reconstruct a multi-dimensional array that had been serialized to a file. For example,

``` js
// f is some interesting function we wish to study

for(var i=0; i<10; i++) {
  for(var j=0; j<10; j++) {
    for(var k=0; k<10; k++) {
      fs.appendLine("myfile.txt", f(i, j, k));
    }
  }
}
```

Later,

``` js
// loadMyArray is some function to read a file into an array

var flat = loadMyArray("myfile.txt");
var array = flat.reshape(10, 10, 10);
// Array is now a multidimensional array
```

Came in very handy to reshape data so it could be plotted by matlab.

**Comment by josdejong** [comment]
Thanks for the examples guys. `reshape` can indeed be a handy construction utility.

@patgrasso it would be great if you want to implement `reshape`, thanks for the offer!

**Comment by iahu** [comment]
`reshape` is useful. for example, to transform a 1d image pixel data array to a M*N shape matrix.

```ts
const pixelData = [255, 255, 255, 255, ....]; // a 1d Buffer array
reshape(pixelData, 20, 20);
```
